server/src/main/java/org/apache/druid/metadata/SqlSegmentsMetadataManager.java

```diff
@@ -1088,6 +1123,174 @@ public DataSegment map(int index, ResultSet r, StatementContext ctx) throws SQLE
+    if (lastSchemaIdPrePoll == null) {
+      log.info("Executing full schema refresh.");
+      schemaPollQuery =
+          StringUtils.format(
+              "SELECT id, payload FROM %s WHERE version = '%s'",
+              getSegmentSchemaTable(),
+              CentralizedDatasourceSchemaConfig.SCHEMA_VERSION
+          );
+    } else {
+      schemaPollQuery = StringUtils.format(
+          "SELECT id, payload FROM %s WHERE version = '%s' AND id > %s",
+          getSegmentSchemaTable(),
+          CentralizedDatasourceSchemaConfig.SCHEMA_VERSION,
+          lastSchemaIdPrePoll
+      );
+    }
+    String finalSchemaPollQuery = schemaPollQuery;
+
+    final AtomicReference<Long> maxPolledId = new AtomicReference<>();
+    maxPolledId.set(lastSchemaIdPrePoll);
+
+    connector.inReadOnlyTransaction(new TransactionCallback<Object>()
+    {
+      @Override
+      public Object inTransaction(Handle handle, TransactionStatus status)
+      {
+        return handle.createQuery(finalSchemaPollQuery)
+                     .setFetchSize(connector.getStreamingFetchSize())
+                     .map(new ResultSetMapper<Void>()
+                     {
+                       @Override
+                       public Void map(int index, ResultSet r, StatementContext ctx) throws SQLException
+                       {
+                         try {
+                           long id = r.getLong("id");
+
+                           if (maxPolledId.get() == null || id > maxPolledId.get()) {
+                             maxPolledId.set(id);
+                           }
+
+                           schemaMap.put(
+                               r.getLong("id"),
+                               jsonMapper.readValue(r.getBytes("payload"), SchemaPayload.class)
+                           );
+                         }
+                         catch (IOException e) {
+                           log.makeAlert(e, "Failed to read schema from db.").emit();
+                         }
+                         return null;
+                       }
+                     }).list();
+      }
+    });
+
+    log.debug("SchemaMap polled from the database is [%s]", schemaMap);
+
+    if (lastSchemaIdPrePoll == null) {
+      // full refresh
+      segmentSchemaCache.updateFinalizedSegmentSchemaReference(schemaMap);
+    } else {
+      // delta update
+      schemaMap.forEach(segmentSchemaCache::addFinalizedSegmentSchema);
+    }
+    segmentSchemaCache.updateFinalizedSegmentStatsReference(segmentStats);
+    segmentSchemaCache.resetInTransitSMQResultPublishedOnDBPoll();
+
+    if (lastSchemaIdPrePoll == null) {
+      segmentSchemaCache.setInitialized();
+    }
+
+    // It is possible that the schema cleanup duty resets the {@code latestSchemaId}.
+    // In that case, we shouldn't update this value with the latest polled schemaId.
+    latestSchemaId.compareAndSet(lastSchemaIdPrePoll, maxPolledId.get());
+
+    segmentSchemaCache.emitStats();
```

Comment:
We are calling reset and then emitting stats. 
Lets get the numbers in the transaction block and use that to emit stats. 
So we would need the segmentSchema to expose a method for giving stat object.